### ISSUE TITLE
pre-Controls 4/5 builds error on pilot select screen if last pilot is from newer build

### ISSUE THREAD

**Issue by notimaginative** [scp-fs2open/fs2open.github.com#3902]
When initializing the pilot select screen it checks the last used pilot. If that pilot has the new control setup (controls 4/5) then it will crash older builds while attempting to read the controls section. The only way around the crash is to remove or change the "LastPlayer" entry in the ini file.

Though I do not consider this a bug in controls 4/5, but a fundamental issue with the pilot code in general, the new controls code could be altered to handle this better.  Notably, if it used a new controls section ("controls_new", or something) and either discarded the old "controls" section after conversion or just saved it to be written back later for compatibility sake. Old builds would still discard the new controls section, but with that info now in a separate file it isn't that big of a loss.

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff5d0b537 in __GI_abort () at abort.c:79
#2  0x00000000015ae4f7 in os::dialogs::Error(char const*) (text=0x4dbfb00 "Error: JSON reading requires a value with name 'controls' but there is no such value!\nFile: JSONFileHandler.cpp\nLine: 209\n") at /home/taylor/Working/src/code/fs2open/code/osapi/dialogs.cpp:293
#3  0x00000000015af461 in os::dialogs::Error(char const*, int, char const*, ...) (filename=0x1e68801 "JSONFileHandler.cpp", line=209, format=0x1e68b8d "JSON reading requires a value with name '%s' but there is no such value!") at /home/taylor/Working/src/code/fs2open/code/osapi/dialogs.cpp:285
#4  0x000000000191e4c8 in pilot::JSONFileHandler::ensureExists(char const*) (this=0x3395b00, name=0x1e3705b "controls") at /home/taylor/Working/src/code/fs2open/code/pilotfile/JSONFileHandler.cpp:209
#5  0x000000000191eba9 in pilot::JSONFileHandler::startArrayRead(char const*, bool) (this=0x3395b00, name=0x1e3705b "controls") at /home/taylor/Working/src/code/fs2open/code/pilotfile/JSONFileHandler.cpp:310
#6  0x000000000166aa61 in pilotfile::plr_read_controls() (this=0x29aac18 <Pilot>) at /home/taylor/Working/src/code/fs2open/code/pilotfile/plr.cpp:657
#7  0x000000000166c66b in pilotfile::load_player(char const*, player*, bool) (this=0x29aac18 <Pilot>, callsign=0x248ff90 <Player_select_last_pilot> "taylor", _p=0x29ac1f0 <Players>, force_binary=false) at /home/taylor/Working/src/code/fs2open/code/pilotfile/plr.cpp:992
#8  0x000000000138000b in player_select_get_last_pilot_info() () at /home/taylor/Working/src/code/fs2open/code/menuui/playermenu.cpp:843
#9  0x000000000137feb9 in player_select_init() () at /home/taylor/Working/src/code/fs2open/code/menuui/playermenu.cpp:329
#10 0x0000000000f88a19 in game_enter_state(int, int) (old_state=0, new_state=37) at /home/taylor/Working/src/code/fs2open/freespace2/freespace.cpp:6007
#11 0x00000000010c5921 in gameseq_set_state(int, int) (new_state=37, override=0) at /home/taylor/Working/src/code/fs2open/code/gamesequence/gamesequence.cpp:286
#12 0x0000000000f86691 in game_process_event(int, int) (current_state=0, event=49) at /home/taylor/Working/src/code/fs2open/freespace2/freespace.cpp:4943
#13 0x00000000010c5e2e in gameseq_process_events() () at /home/taylor/Working/src/code/fs2open/code/gamesequence/gamesequence.cpp:379
#14 0x0000000000f8a2ce in game_main(int, char**) (argc=3, argv=0x7fffffffde38) at /home/taylor/Working/src/code/fs2open/freespace2/freespace.cpp:6604
#15 0x0000000000f8b5d1 in main(int, char**) (argc=3, argv=0x7fffffffde38) at /home/taylor/Working/src/code/fs2open/freespace2/freespace.cpp:7666
```

**Comment by z64555** [comment]
Why was that an error in the first place? Even if the controls were empty or corrupted, you could still play the game with defaults.

At any rate, what are the reasons here to make the new pilotfiles backwards compatible?

**Comment by notimaginative** [comment]
The issue is that when someone uses 22.0 their pilot is upgraded to the new controls setup, and if they go back to 21.4 or an earlier nightly for some reason it will error at the start since it can't read the last used pilot file, which is does automatically. So you can't get into the game at all unless you edit the ini file or the json. This makes testing pre-Controls 4/5 builds difficult (in the best case) once a pilot has been updated.

**Comment by z64555** [comment]
That's reason enough, I'll get to it over the weekend.

**Comment by z64555** [comment]
Please toss me a copy of the pilot .json that causes this issue

**Comment by notimaginative** [comment]
It's just a brand new pilot created by a current nightly. The same pilot opened with a pre-Controls 4/5 nightly errors at startup in every attempt so far. Opening 21.4 appears to do one of three things: error at startup (most often for me), error when trying to load the pilot, or not show the pilot at all. I haven't figured out the reason for why the behavior isn't consistent.

**Comment by z64555** [comment]
gimme anyway, plz

**Comment by notimaginative** [comment]
[test.json.gz](https://github.com/scp-fs2open/fs2open.github.com/files/7833842/test.json.gz)

**Comment by z64555** [comment]
Alright, that confirms that.

So, the course of action I settled on before having a massive headache stop me, was to have the default bindings written to the player .json (and .csg, if needed) for the sole objective of preventing a crash.  I've already decided beforehand that I'm not going to support full forward compatibility to allow new v4 .plr's and v7 .csg's be usable by old versions of FSO, but the switching back and forth between versions is almost unavoidable at this time because of the way that FSO handles playerfiles at startup.
